member 도메인 리팩토링

vectorch9 opened the pull request:
<!--
✅ Resolve: #이슈번호 형태로 입력해 주세요.
ex) Resolve: #123
-->
## 📌Linked Issues
- Resolve: #159 

<!--
✅ 변경 사항을 자세히 알려주세요. (why -> what -> how)
-->
## ✏Change Details
### member 도메인에 컨벤션 적용
- 지난번 유저 인증을 제외한 부분에 컨벤션 적용
### AuthUtil의 메서드 일부 수정
- static으로 지정 되어있던 `getLoginMemberIdOrNull` 메서드를 일반 메서드로 변경
- 이에 따라 연관된 API 수정
### Member 도메인 리팩토링
- member 도메인에 AuthUtil을 적용해 중복 코드 제거
- MemberNotFoundException을 EntityNotFoundException으로 변경
- querydsl 관련 리팩토링
  - 중복되는 서브쿼리를 메서드로 분리
  - 의미 없는 매개변수(loginUserId) 삭제
  - loginedUserId를 loginUserId로 변경
  - 차단함/차단당함 여부를 Querydsl로 구현 => 기존 2개의 쿼리를 1개의 쿼리로 줄임
  - MemberPost와 관련된 유사한 기능의 메서드를 하나로 통합(e.g. getRecent15MemberPostDTOs와 getMemberPostDTOs를 통합)
  - 하나의 메서드가 하나의 기능만 하도록 분리
  - 차단여부를 확인하는 메서드(checkBlock)를 제거 => 이제 다른 메서드 실행시 자동적으로 수행됨 
- DTO 수정
  - MemberPostDTO내 postImageUrl(String)을 PostImageDTO로 변경
- 유저 관련 서비스 리팩토링
  - MemberPostService에서 null이 아닌 []을 반환하도록 수정
  - MemberPostService내 중복코드를 setMemberPostImageDTOs로 분리
  - setHasStory를 repository가 아닌 서비스 층에서 사용하도록 변경

<!--
✅ 추가로 전달할 내용이 있다면 적어주세요.
-->
## 💬Comment
- Empty Page, MemberPostDTO의 content 변경은 배포후 프론트분들께 전달하겠습니다.
- 저도 IntelliJ로 넘어 왔습니다.. 😳 Eclipse 기반 formatter가 좀 더 엄격한 느낌이 있어 vscode로 foramtting하면 마음에 안드는 개행이 생기더라구요..
<!--
✅ 참고한 사이트가 있다면 공유해주세요.
-->
## 📑References

## ✅Check List
- [x] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?

vectorch9 committed:
Style: member 도메인에 컨벤션 적용

vectorch9 committed:
Refactor: AuthUtil의 static 메서드를 일반 메서드로 수정

- getLoginMemberIdOrNull 메서드를 일반 메서드로 변경
    - 이에따라 해당 메서드를 호출하던 코드 수정

vectorch9 committed:
Refactor: member도메인에 AuthUtil 적용, 중복제거

- member 도메인에 AuthUtil을 적용해 중복 코드 제거
- MemberNotFoundException을 EntityNotFoundException으로 변경

vectorch9 committed:
Refactor: 멤버 API 리팩토링

- querydsl 관련 리팩토링
  - 중복되는 서브쿼리를 메서드로 분리
  - 의미 없는 매개변수(loginUserId) 삭제
  - loginedUserId를 loginUserId로 변경
  - 차단함/차단당함 여부를 Querydsl로 구현 => 기존 2개의 쿼리를 1개의 쿼리로 줄임
  - MemberPost와 관련된 유사한 기능의 메서드를 하나로 통합(e.g. getRecent15MemberPostDTOs와 getMemberPostDTOs를 통합)
  - 하나의 메서드가 하나의 기능만 하도록 분리
  - 차단여부를 확인하는 메서드(checkBlock)를 제거 => 이제 다른 메서드 실행시 자동적으로 수행됨
- 컨벤션관련 수정
  - 이전의 커밋에서 수정한 컨벤션 중 적용되지 않던 사항이 있어 수정
  - EntityNotFoundException관련 수정
- DTO 수정
  - MemberPostDTO내 postImageUrl(String)을 PostImageDTO로 변경
- 유저 관련 서비스 리팩토링
  - MemberPostService에서 null이 아닌 []을 반환하도록 수정
  - MemberPostService내 중복코드를
  setMemberPostImageDTOs로 분리
  - setHasStory를 repository가 아닌 서비스 층에서 사용하도록 변경

vectorch9 committed:
Fix: 잘못된 Indent 수정

seonpilKim reviewed (commented):
commit message 작성하실 때, `Fix`는 버그 수정인 경우에만 사용하시면 좋을 것 같아요.
들여쓰기, 오타, 네이밍 수정 등 비즈니스 로직과 상관 없는 코드 수정은 `Style`을 사용하시면 됩니다.

그리고 클래스 멤버변수를 private으로 캡슐화하지 않은 파일들이 몇 개 보이던데, 이 부분들도 추가해 주시면 좋을 것 같아요.

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/controller/MemberPostController.java:
```diff
@@ -32,28 +32,28 @@ public class MemberPostController {
 
```
endpoint 중에 `/accounts`가 중복되는데, 이거를 클래스단위로 `@RequestMapping("/accounts")` 빼면 유지보수시에 더 편리할 것 같아요!

vectorch9 replied:
1. 넵 Swagger 관련은 적용하면 좋을 것 같습니다. 
2. Effective C++에서 `const `관련 글을 읽은 뒤로 생각나는 대로 `final`을 붙이려고 하는데 맨날 까먹네요.. 
3. 저도 Class 블록 시작과 끝에 빈줄이 있는게 가독성이 좋더라구요. 메서드는 아마 앞뒤의 빈줄은 다 제거 했을텐데 클래스는 통일되어 있지 않아서 전부 빈줄을 삽입해 통일하겠습니다.

seonpilKim replied:
@bluetifulc 
1. 네네 그럼 저도 이대로 작업할게요.
2. 인텔리제이 쓰시면 변수 추출 단축키로 final 자동으로 추가해 줍니다 ㅎㅎ
3. 메소드에서도 몇 개 발견했었어요! 
[MemberRepositoryQuerydslImpl.java](https://github.com/Instagram-Clone-Coding/Spring_instagram-clone/pull/160/files#diff-ee277f09ea95aef0b9fb4c94d5ff92bff1e4a2e14df732b586cb3766c1f5444e) 여기 말고도 있었던 것 같은데 기억이 안나네요

seonpilKim replied:
아 그리고 SwaggerConfig는 제가 다른 브런치에서 수정해서 PR 올릴게요

vectorch9 replied:
넵 그럼 ApiResponse는 다음 브랜치에 적용하겠습니다

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/controller/MemberPostController.java:
```diff
@@ -32,28 +32,28 @@ public class MemberPostController {
 
```
#### Bean Validation 관련
이전부터 Bean Validation 관련 의문점이 몇 개 있어서 자세히 찾아봤는데, 저희가 사용하는 방식에 문제가 있는 것 같아요.

- `@RequestParam`, `@PathVariable` 대상으로는 `@Validated` or `@Valid`이 적용되지 않음
`@ModelAttribute`, `@RequestBody` 이런 어노테이션은 ArgumentResolver의 바인딩 과정에 Validation도 포함되지만, `@RequestParam`, `@PathVariable`은 Validation 과정 없이 바로 Primitive type으로 변환된다고 합니다.
    > `@Validated`를 클래스 단위에 추가하면, `AOP`방식으로 동작해서 Validation 과정이 추가된다고 합니다.
따라서, 클래스 단위에 `@Validated`를 추가하고, 파라미터에서는 빼면 될 것 같아요.
    - 추가로, `@RequestParam` 옵션 중 required = true가 default라서, `@RequestParam`, `@PathVariable` 사용 시에는 `@NotNull`이 불필요한 것 같아요. 

- `@ModelAttribute`, `@RequestBody` 파라미터 검증 시, `@Validated` 대신 `@Valid` 사용으로 바꾸는 건 어떨까요?
메소드 파라미터에 적용하는 경우, 두 어노테이션의 차이로 `Validation groups` 가 있는데, 저희는 이 기능을 사용하지 않으니까, 해당 기능이 없는 `@Valid`를 사용하는 게 좋을 것 같아요.

#### References
- https://medium.com/javarevisited/are-you-using-valid-and-validated-annotations-wrong-b4a35ac1bca4
- https://beaniejoy.tistory.com/72

vectorch9 replied:
- Class에 `@Validated`를 추가하면 `RequestParam`, `PathVariable`은 알아서 AOP가 적용되고, `RequestBody`, `ModelAttribute`는 `@Validated`나 `@Valid`를 Class와 관계없이 붙여주어야 한다고 이해했는데 맞나요?
- `Valid`와 `Validated` 차이는 프로젝트 초반에 궁금해서 찾아봤었는데, 언젠간 groups을 쓸 수도 있지 않을까 해서 그냥 뒀는데 결국 안쓰게 됐네요.. 그 동안 groups를 쓴 사례도 본적이 없어서 Valid로 바꿔도 좋을 것 같습니다

seonpilKim replied:
네 맞아요. 추가로 Controller가 아닌 Service나 Repository에서도 AOP 방식으로 Bean Validation를 적용할 수 있어요. 클래스 단위에 `@Validated` + 파라미터(dto)에 `@Valid` -> 쓰일 진 모르겠지만..

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/dto/MiniProfileResponse.java:
```diff
@@ -19,8 +20,6 @@
 
 @ApiModel("유저 미니프로필 응답 모델")
```
Response가 스웨거 모델에 추가가 안되니까, 앞으로 이 어노테이션은 추가하지 않아도 될 것 같아요.

vectorch9 replied:
저희는 ResultResponse로 wrapping 후 반환을 해서 swagger에 표시가 안되는데, 현업에서는 그럼 Wrapper 없이 반환하는 걸까요?

seonpilKim replied:
@bluetifulc 
저도 잘 모르겠네요😅

seonpilKim reviewed (commented):
#### 스웨거 어노테이션 관련
- 컨트롤러에서 인증이 필요하지 않은 메소드마다 어노테이션을 추가하는 게 번거롭고 불필요한 작업인 것 같다는 생각이 들었어요.
    - SwaggerConfig에서 requried를 false로 설정하면, 따로 어노테이션을 추가하지 않아도 돼서 편할 것 같은데 어떠신가요?
    - 그리고 각 컨트롤러의 메소드마다 `@ApiResponse(code = xxx, message = xxx)`를 추가하면, 저번에 수화님이 부탁하신 것도 해결되고, 인증이 필요한지 여부도 알 수 있어서 좋을 것 같아요.
    - ![image](https://user-images.githubusercontent.com/68049320/167382161-4eaf2f02-ea9b-4fab-b14a-bc6f3e7410de.png)
        - message: `비즈니스 코드 - 메시지` 
        - API를 호출했을 때 발생할 수 있는 결과들에 대한 응답만 추가하면 될 것 같아요.
    - 모든 API에 공통으로 발생하는 응답은 SwaggerConfig에 아래와 같이 추가해서 사용하면 됩니다.
        - ![image](https://user-images.githubusercontent.com/68049320/167382314-bfce15e3-73ec-42fa-9268-c0d36e6c6fa1.png)
- 값이 변경될 일이 없는 변수에는 final을 명시하면 좋을 것 같아요. 
    - 특히 List도 add 메소드로 원소를 추가하는 형태라, List 자체를 바꾸지 않는 이상 final 명시하는게 바람직할 것 같아요.

#### 컨벤션 관련
- 저번에 카톡으로 빈 줄 관련해서 얘기했던 것 같은데, 다시 보니까 `파일의 마지막에는 새 줄`을 추가하라는 컨벤션이 있었어요. 인텔리제이에서 코드 정렬할 때도 보면 항상 마지막에 새 줄이 추가되더라구요.
- 컨벤션에 명시되어 있지는 않지만, 라이브러리 코드를 보면 공통점이 있는데, 이 부분도 같이 적용하면 가독성 측면에서 더 좋을 것 같아요.
    - Class 블록 시작과 끝에 새 줄 추가
    - method 블록에는 처음과 끝에 새 줄 추가하지 않음

vectorch9 committed:
Refactor: Member 도메인 Controller 리팩토링
- final 적용할 수 있는 부분에 적용
- return 형식 변경: result 객체를 따로 만드는 방식에서 ResponseEntity의 ok함수를 이용하는 방식으로 변경
- Bean Validation
  - Controller class에 @Validated 추가
  - 의미없는 @Validated 삭제
  - DTO의 검증은 @Valid로 통일
- 중복되는 코드 메서드로 분리(refreshToken 쿠키 생성)
- 저장된 게시물 API의 endpoint 변경(의미 없는 username 삭제)

vectorch9 committed:
Refactor: member 도메인 서비스 리팩토링
- 컨벤션 추가적용
  - final 키워드를 사용할 수 있는 곳에 추가
  - 빈 줄 컨벤션 적용
- Service 리팩토링
  - 오타수정 reisuue -> reissue
  - 의미 없는 if문 삭제(return과 연관되어 한줄로 줄 일 수 있는 경우 개선)
  - MemberPostService의 page 파라미터 0과 1을 구분하는 로직 삭제 -> Controller단 에서 Bean Validation에 의해 검증되도록 수정
  - 반복되는 상수(magic number)를 static final 변수로 수정

vectorch9 committed:
Style: DTO 빈줄 컨벤션 적용

- 빈줄 컨벤션 적용
- private 누락된 일부 DTO 수정
- LoginedDevicesDTO를 LoginDevicesDTO로 수정

vectorch9 committed:
Style: Entity, Repository 빈줄 컨벤션 적용
- 빈줄 컨벤션 적용
- final 추가 가능한 부분에 추가

vectorch9 committed:
Style: member 도메인 Exception에 빈줄 컨벤션 적용

vectorch9 committed:
Fix: 비밀번호 재설정 메일 내 username이 dlwlrma으로 고정되던 문제 수정
- 커밋명의 문제 수정
- 메일 내 a태그의 href에 중괄호가 포함되던 문제 수정

vectorch9 committed:
Refactor: 비밀번호 수정 시 중복 검사 로직 추가
- 기존의 비밀번호와 일치하면 예외를 발생시키도록 수정

vectorch9 commented:
## 추가 변경사항
### 일부 endpoint 변경
- MemberController의 endpoint prefix 통일(accounts)
  - 이에 따라 기존 `/menu/profile`의 메뉴 멤버 프로필 조회 API가 `/accounts/profile`로 변경됨
- 로그인한 기기 조회 API의 endpoint 변경 `/accounts/logined/device` -> `/accounts/login/device`
### 컨벤션 추가 적용 및 전체적인 수정사항
- 빈 줄 컨벤션 member 도메인에 적용
  - 클래스의 시작과 끝에 빈줄 추가
  - 파일의 끝에 빈줄 추가
  - 메서드 시작과 끝의 빈줄 제거
- final을 적용가능한 부분에 적용
### 서비스 리팩토링
- 오타수정 reisuue -> reissue
- 의미 없는 if문 삭제(return과 연관되어 한줄로 줄일 수 있는 경우 개선)
- page 파라미터 0과 1을 구분하는 로직 삭제 -> Controller단 에서 Bean Validation에 의해 검증되도록 수정
- 반복되는 상수(magic number)를 static final 변수로 수정
### 비밀번호 변경에 중복 검사 추가
- 비밀번호 변경 시 기존의 비밀번호와 같다면 예외 발생
### 컨트롤러 리팩토링
- 컨트롤러의 return 형식 변경: result 객체를 따로 만드는 방식에서 `ResponseEntity`의 `ok` 메서드를 이용하는 방식으로 변경
- Bean Validation
  - Controller class에 `@Validated` 추가
  - 의미없는 `@Validated` 삭제(RequestParam, PathVariable)
  - DTO의 검증은 `@Valid`로 통일
- 중복되는 코드 메서드로 분리(refreshToken 쿠키 생성)
- 저장된 게시물 API의 endpoint 변경(의미 없는 username 삭제)
- 페이지 조회 API에 `@MIN(1)` 추가
### 버그수정
 - 비밀번호 재설정 메일 내 일부분에서 username이 고정되던 문제 수정
- 메일 내 a태그의 href에 중괄호가 포함되던 문제 수정

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/controller/BlockController.java:
```diff
@@ -21,30 +19,27 @@
-	public ResponseEntity<ResultResponse> block(
-			@PathVariable("blockMemberUsername") @Validated @NotBlank(message = "username이 필요합니다") String blockMemberUsername) {
-		boolean success = blockService.block(blockMemberUsername);
+	public ResponseEntity<ResultResponse> block(@PathVariable("blockMemberUsername") String blockMemberUsername) {
```
`@PathVariable` 옵션 default로 requried = true가 적용되어서 `@NotNull`은 불필요하지만, `@NotBlank`는 추가해도 좋을 것 같아요.

vectorch9 replied:
음.. DB에서 "해당 username이 존재하지 않습니다"로 걸러지기 때문에 뺐는데 DB단 까지 가는거보단 컨트롤러에서 처리하는게 낫겠네요. 이 부분도 수정하겠습니다.

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/controller/MemberController.java:
```diff
@@ -17,95 +21,87 @@
 @Api(tags = "멤버 API")
 @RestController
 @RequiredArgsConstructor
+@RequestMapping("accounts")
```
앞에 `/` 안 붙여도 될까요?

vectorch9 replied:
안붙여도 정상작동해서 못보고 있었네요. 추가하겠습니다

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/service/MemberAuthService.java:
```diff
@@ -132,11 +132,12 @@ public Optional<JwtDto> reissue(String refreshTokenString) {
 			throw new AccountMismatchException();
 		}
+		if (updatePasswordRequest.getOldPassword().equals(updatePasswordRequest.getNewPassword())) {
+			throw new EntityAlreadyExistException(PASSWORD_ALREADY_EXIST);
```
Password는 엔티티의 필드라서 EntityAlreadyExistException 보다는 다른 예외를 사용하는 건 어떨까요?

vectorch9 replied:
수정하겠습니다.

seonpilKim reviewed (commented):
엇 근데 충돌나네요

vectorch9 committed:
Merge branch 'develop' into Refactor/Member
